### ISSUE TITLE
Weighted Sample Leads to Iteration Fails

### ISSUE THREAD

**Issue by WXFMAV** [Microsoft/LightGBM#939]
## Environment info
Operating System: MacOs
CPU:
C++/Python/R version: Python
## Error Summary

The algorithm simply can not iterate when I applying weighted sample technique, this problem occurs only when I set 'objective' as 'regression_l1' and valid the weighted sample algorithm. Additionaly, the weight of each sample varies from 0~100000  in our project . 
I wonder if I have missed some important params to set before I refer to the weighted sample feature.

## Params set

params = {
    'task': 'train',
    'boosting_type': 'gbdt',
    'objective': **'regression_l1'**,
    'metric': {'mae','mse'},
    'num_leaves': 63,
    'min_data_in_leaf': 20,
    'learning_rate': 0.01,
    'feature_fraction': 0.8,
    'bagging_fraction': 0.7,
    'bagging_freq': 5,
    'verbose': 0
}

## Error Message:

 lgb_train = lgb.Dataset(X_train, label=y_train, **weight = w_train**)
 lgb_eval = lgb.Dataset(X_test, label=y_test, **weight = w_test**, reference=lgb_train)

 fails to iterate

[1]	valid_0's l1: **1.54593**	valid_0's l2: 4.19559
[2]	valid_0's l1: **1.54593**	valid_0's l2: 4.19559
[3]	valid_0's l1: 1.54593	valid_0's l2: 4.19559
[4]	valid_0's l1: 1.54593	valid_0's l2: 4.19559
[98]	valid_0's l1: 1.54593	valid_0's l2: 4.19558
[99]	valid_0's l1: 1.54593	valid_0's l2: 4.19558
[100]	valid_0's l1: 1.54593	valid_0's l2: 4.19558
[101]	valid_0's l1: 1.54593	valid_0's l2: 4.19558
[242]	valid_0's l1: 1.54593	valid_0's l2: 4.19557
[243]	valid_0's l1: 1.54593	valid_0's l2: 4.19557
[244]	valid_0's l1: 1.54593	valid_0's l2: 4.19557
[263]	valid_0's l1: 1.54593	valid_0's l2: 4.19557
[265]	valid_0's l1: 1.54593	valid_0's l2: 4.19557
[266]	valid_0's l1: **1.54593**	valid_0's l2: 4.19557
[267]	valid_0's l1: **1.54593**	valid_0's l2: 4.19557

## Code before applying weighted sample works well:
 lgb_train = lgb.Dataset(X_train, label=y_train)
 lgb_eval = lgb.Dataset(X_test, label=y_test, reference=lgb_train)

iterates well

[1]	valid_0's l1: **2.03321**	valid_0's l2: 33.2134
[2]	valid_0's l1: **2.00531**	valid_0's l2: 32.9768
[6]	valid_0's l1: **1.89996**	valid_0's l2: 32.0771
[305]	valid_0's l1: 0.985381	valid_0's l2: 21.7331
[314]	valid_0's l1: 0.981737	valid_0's l2: 21.6612
[315]	valid_0's l1: 0.981381	valid_0's l2: 21.6549
[350]	valid_0's l1: 0.968002	valid_0's l2: 21.3568
[351]	valid_0's l1: **0.967573**	valid_0's l2: 21.3479
[352]	valid_0's l1: **0.967215**	valid_0's l2: 21.3411

**Comment by guolinke** [comment]
@henry0312

**Comment by henry0312** [comment]
I'll check next week, sorry.

**Comment by henry0312** [comment]
@WXFMAV can you give me reproduceable code?

**Comment by WXFMAV** [comment]
It was impossible to upload the huge train data files but the Codes that leads to errors can be show as below
```
# coding: utf-8
# pylint: disable = invalid-name, C0111
import json
import lightgbm as lgb
import pandas as pd
from sklearn.metrics import mean_squared_error

df_train = read_sparse('train.txt');
df_test = read_sparse('test.txt');

y_train = df_train['label'].values
y_test = df_test['label'].values
X_train = df_train.drop(['label','type','base','item_id'], axis=1).values
X_test = df_test.drop(['label','type','base','item_id'], axis=1).values
w_test = df_test['base'].values
w_train = df_train['base'].values

feature_name = df_train.drop(['label','type','base','item_id'],axis=1).columns.values.tolist()
print feature_name

lgb_train = lgb.Dataset(X_train, label=y_train, weight = w_train)
lgb_eval = lgb.Dataset(X_test, label=y_test, weight = w_test, reference=lgb_train)

params = {
    'task': 'train',
    'boosting_type': 'gbdt',
    'objective': 'regression_l1',
    'metric': {'mae','mse'},
    'num_leaves': 63,
    'min_data_in_leaf': 20,
    'learning_rate': 0.01,
    'feature_fraction': 0.8,
    'bagging_fraction': 0.7,
    'bagging_freq': 5,
    'verbose': 0
}

num_tree = 500

print('Start training...')
# train
gbm = lgb.train(params,
                lgb_train,
                num_boost_round=num_tree,
                feature_name = feature_name,
                valid_sets=lgb_eval)
#                early_stopping_rounds=5)

```

outputs:

```
[1]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[2]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[3]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[4]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[5]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[6]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[7]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[8]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[9]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[10]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[11]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[12]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[13]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[14]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[15]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[16]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[17]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[18]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[19]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[20]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[21]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[22]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[23]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[24]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[25]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[26]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[27]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[28]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[29]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[30]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[31]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[32]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[33]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[34]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[35]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[36]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[37]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[38]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[39]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[40]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[41]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[42]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[43]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[44]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[45]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[46]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[47]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[48]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[49]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
[50]	valid_0's l1: 1.23869	valid_0's l2: 2.89915
```

while the head 20 records of **w_train** is 
```
0     919016
1     757362
2     649980
3     503680
4     502539
5     478614
6     448869
7     412322
8     392391
9     389558
10    381897
11    375394
12    360402
13    358164
14    326048
15    320238
16    304377
17    303570
18    299455
19    297538
``` 
it seems that the error can be avoid when the **w_train** is small enough.

**Comment by WXFMAV** [comment]
@henry0312  The codes works well if only we set 
```
'objective': 'regression_l1'
```
into 
```
'objective': 'regression_l2'
```

**Comment by henry0312** [comment]
When I have free time, I'll check this issue, sorry.

**Comment by guolinke** [comment]
@henry0312 any updates ?

**Comment by guolinke** [comment]
move the discussion to #979
